src/priceClient/adapters/defiLlama.ts

```diff
@@ -0,0 +1,81 @@
+import assert from "assert";
+import { PriceFeedAdapter, TokenPrice } from "../priceClient";
+import { BaseHTTPAdapter } from "./baseAdapter";
+
+type DefiLlamaTokenPrice = {
+  price: number;
+  symbol: string;
+  timestamp: number;
+  confidence: number;
+};
+
+type DefiLlamaPriceResponse = {
+  coins: {
+    [key: string]: DefiLlamaTokenPrice;
+  };
+};
+
+const defaultTimeout = 5000; // mS
+
+export class PriceFeed extends BaseHTTPAdapter implements PriceFeedAdapter {
+  private _minConfidence: number;
+
+  constructor(name = "DefiLlama", { timeout, minConfidence = 0.9 }: { timeout?: number; minConfidence?: number }) {
+    super(name, "coins.llama.fi", { timeout: timeout ?? defaultTimeout });
+    this._minConfidence = minConfidence;
+  }
+
+  get minConfidence(): number {
+    return this._minConfidence;
+  }
+
+  set minConfidence(minConfidence: number) {
+    assert(minConfidence >= 0.0 && minConfidence <= 1.0);
+    this._minConfidence = minConfidence;
+  }
+
+  async getPriceByAddress(address: string, currency = "usd"): Promise<TokenPrice> {
+    const price = await this.getPricesByAddress([address], currency);
+    return price[0];
+  }
+
+  async getPricesByAddress(addresses: string[], currency = "usd"): Promise<TokenPrice[]> {
+    if (currency != "usd") throw new Error(`Currency ${currency} not supported by DefiLlama`);
+
+    const path = "prices/current/" + addresses.map((address) => `ethereum:${address}`).join();
+    const tokenPrices: unknown = await this.query(path, {});
+    if (!this.validateResponse(tokenPrices))
+      throw new Error(`Unexpected ${this.name} response: ${JSON.stringify(tokenPrices)}`);
+
+    return addresses.map((addr: string) => {
+      const tokenPrice = tokenPrices.coins[`ethereum:${addr.toLowerCase()}`];
+      if (tokenPrice === undefined) throw new Error(`Token ${addr} missing from ${this.name} response`);
+
+      if (tokenPrice.confidence < this.minConfidence)
+        throw new Error(`Token ${addr} has low confidence score (${tokenPrice.confidence})`);
+
+      return { address: addr, price: tokenPrice.price, timestamp: tokenPrice.timestamp };
+    });
+  }
+
+  private validateResponse(response: unknown): response is DefiLlamaPriceResponse {
+    if (response === null || typeof response !== "object") return false;
+
+    const coins: unknown = (response as Record<string, unknown>).coins;
+    if (coins === null || typeof coins !== "object") return false;
+
+    return Object.entries(coins as object).every(([address, tokenPrice]) => {
+      // prettier-ignore
+      return (
+        /[a-z]+:0x[0-9a-fA-F]{40}/.exec(address) !== undefined
```

Comment:
I feel like there shoud be a better way to validate an HTTP response. But I see this is consistent with the coingecko adapter. Have you heard of [yup](https://github.com/jquense/yup)? 

@dohaki has used this in the merkle distributor scripts [here](https://github.com/across-protocol/merkle-distributor/pull/1) you should look into it!

Reply:
Yeah, we use yup and superstruct to validate typescript types in different places. It would be great to standardize around one of them! Do people have strong preferences between the two?

@pxrl I do think we generally should prefer using one of those libs over manual verification.

Reply:
Just summing up some thoughts after spending a bit of time investigating this today.

TLDR: I can see some advantages to using a dedicated package for validation, but I haven't yet found a way for them to efficiently validate data structures with a mixture of fixed and dynamic keys.

See the bottom of this message for an overview of the data structures; dynamic keys are used for Eth addresses and currencies. This seems to be a case that isn't handled handle "out of the box" for these packages; there are threads where others have proposed solutions/workarounds, but they seem a bit hackish and require a surprising amount of boilerplate. and (to me) it becomes pretty difficult to understand what's going on.

I did find another package called [Joi](https://joi.dev/api/?v=17.6.1#objectpatternpattern-schema-options) that seems to handle this nicely by allowing dynamic keys to be specified as a regex pattern. This would _probably_ be a viable solution, and it seems cool, but involves yet another package with additional dependencies. We also already have superstruct as a dependency in sdk-v2, so if we decided to go for something else then I think it'd be worthwhile rolling it out to the places where we currently use superstruct.

@dohaki If you have any suggestions on how to efficiently handle the mix of fixed and dynamic keys then I'd be interested to hear more! cc @daywiss

For context on the data structures that I am aiming to validate:

CoinGecko:
```
{
  [address: string]: {
     [currency: string]: number;
    last_updated_at: number;
  };
};
```

DefiLlama:
```
{
  [address: string]: {
    symbol: string;
    price: number;
    timestamp: number;
    confidence: number;
  };
};
```